udev_device.c

```diff
@@ -540,6 +545,49 @@ static void set_properties_from_props(struct udev_device *udev_device)
     udev_list_entry_add(&udev_device->properties, "ID_PATH", id, 0);
 }
 
+static int set_properties_from_helper(struct udev_device *udev_device)
+{
+    size_t size = 0;
+    size_t count = 0;
+    struct udev_list_entry *entry;
+    for (entry = udev_device_get_properties_list_entry(udev_device);
+         entry;
+         entry = udev_list_entry_get_next(entry)) {
+
+        const char *name = entry->name;
+        const char *value = entry->value?:"";
+        // '{name}='{value}' '
+        size += strlen(name) + 2 + strlen(value) + 2;
+    }
+
+    char helper_path[] = "/sbin/libudev-zero-rules-helper";
+    char *cmd = malloc(size + sizeof(helper_path));
+    char *ptr = cmd;
+    for (entry = udev_device_get_properties_list_entry(udev_device);
+         entry;
+         entry = udev_list_entry_get_next(entry)) {
+        const char *name = entry->name;
+        const char *value = entry->value?:"";
+        // '{name}='{value}' '
+        size_t max_size = strlen(name) + 2 + strlen(value) + 2;
+        // +1 for \0
+        snprintf(ptr, max_size + 1, "%s='%s' ", name, value);
+        ptr += max_size;
+    }
+    snprintf(ptr, sizeof(helper_path), "%s", helper_path);
+
+    FILE *file = popen(cmd, "r");
```

Comment:
`popen` is rather unsafe due to shell invocation. `posix_spawn` is better alternative, albeit it requires more code.